File: core/src/adapters/nvidia_gpu.rs
use crate::adapters::EnergyAdapter;
use crate::EnergyResult;
use hardware_plugins::{HardwarePlugin, Measurement};

pub struct NvidiaGpuAdapter {
    base: super::BaseAdapter,
}

impl NvidiaGpuAdapter {
    pub fn new(plugin: Box<dyn HardwarePlugin>) -> Self {
        Self {
            base: super::BaseAdapter::new("nvidia_gpu", plugin),
        }
    }
}

impl EnergyAdapter for NvidiaGpuAdapter {
    fn name(&self) -> &str {
        self.base.name()
    }

    fn initialize(&mut self) -> EnergyResult<()> {
        Ok(())
    }

    fn shutdown(&mut self) -> EnergyResult<()> {
        Ok(())
    }

    fn read_measurements(&self) -> EnergyResult<Vec<Measurement>> {
        let measurement = self.base.plugin().get_measurement()?;
        Ok(vec![measurement])
    }
} 

File: src/measurement/src/adapters/nvidia_gpu.cpp
#include "plugin/hardware_plugin.hpp"

namespace codegreen {

class NvidiaGPUAdapter : public HardwarePlugin {
public:
    NvidiaGPUAdapter() = default;
    ~NvidiaGPUAdapter() override = default;

    std::string name() const override { return "NVIDIA GPU"; }
    std::unique_ptr<Measurement> get_measurement() const override { return nullptr; }
    bool init() override { return true; }
    void cleanup() override {}
    bool is_available() const override { return false; }
};

} // namespace codegreen
